### PR TITLE
Integrated extend-oem into cos-customizer

### PR DESCRIPTION
**Pull request by hegao12**
The OEM partition is 12MB in COS image, but we want it to be larger to store read-only user data. So the extend_oem_partition program is integrated into COS-Customizer.
Now users can add a -oem-size=${size} flage to the finish-image-build command to set the OEM partition size. Since the OEM parition needs space for dm-verity hash tree, its size is doubled. The disk size set with -disk-size-gb=${size} must be no smaller than  --oem-size x 2 (rounded up to GB) + 10 to leave enough space for the OEM partition.
If the oem-size is set, startup.sh will create a unit that runs at shutdown time to execute a binary program of extend_oem_partition.go and then reboot the instance. The binary file lives in data/builtin_build_context/ .
Since we allow oem-size like "500M", and the OEM size is rounded up to GB, the extra space will be taken by the stateful partition.

### CODE REVIEW

File: cmd/finish_image_build.go [hunk 1]
```diff
@@ -93,13 +96,34 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+		"can be a number with unit like 10G, 10M, 10K or 10B, "+
+		"or without unit indicating the number of 512B sectors.")
 	flags.IntVar(&f.diskSize, "disk-size-gb", 0, "The disk size to use when creating the image in GB. Value of '0' "+
 		"indicates the default size.")
 	flags.DurationVar(&f.timeout, "timeout", time.Hour, "Timeout value of the image build process. Must be formatted "+
 		"according to Golang's time.Duration string format.")
 }
 
 func (f *FinishImageBuild) validate() error {
+	const IMGSIZE = 10
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: sorry if I missed this in the other PR, but at Google it's idiomatic to name constants the same way that you would name any other variable in Go (so in this case, imgSize). It's not a big deal though.

File: cmd/finish_image_build.go [hunk 2]
```diff
@@ -93,13 +96,34 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+	if f.oemSize != "" {
+		oemSizeGB, err := partutil.ConvertSizeToGBRoundUp(f.oemSize)
+		if err != nil {
+			return fmt.Errorf("invalid format of oem-size: %s, error msg:(%v)", f.oemSize, err)
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: nit: prefer to use %q instead of %s when printing string variable values. It typically makes the result a little easier to read.

File: cmd/finish_image_build.go (hunk 1 above)
Comment (rkolchmeyer) [inline cmd/finish_image_build.go]: Could you add a comment explaining what this constant represents?

File: data/build_image.wf.json [hunk 3]
```diff
@@ -68,6 +71,9 @@
         }
       ]
     },
+    "resize-disk":{ 
+      {{.ResizeDisks}}
+    },
```
Comment (rkolchmeyer) [inline data/build_image.wf.json:77]: If OEMSize == "", will the stateful partition ever be resized up after this step?

File: data/startup.sh [hunk 4]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+[Service]
+Restart=no
+EOF
+
+  /usr/bin/systemctl daemon-reload
+  /usr/bin/systemctl stop systemd-journald.socket
+  /usr/bin/systemctl stop systemd-journald-dev-log.socket
+  /usr/bin/systemctl stop systemd-journald-audit.socket
+  /usr/bin/systemctl stop syslog.socket
+  /usr/bin/systemctl stop systemd-journald.service
```
Comment (rkolchmeyer) [inline data/startup.sh]: nit: using the absolute path of /usr/bin/systemctl is unnecessary, and will break if systemctl moves to somewhere else in the PATH. But /usr/bin/systemctl is pretty stable, so in practice this probably doesn't matter.

File: data/startup.sh [hunk 5]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+}
+
+# this unit runs at shutdown time after everything but /tmp is unmounted
+create_run_after_unmount_unit(){
+  mkdir /tmp/last
```
Comment (rkolchmeyer) [inline data/startup.sh]: Instead of choosing an explicit tmpdir name, it's better to use mktemp -d:

local tmpdir
tmpdir="$(mktemp -d)"

File: data/startup.sh [hunk 6]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+  mount -t tmpfs tmpfslast /tmp/last
```
Comment (rkolchmeyer) [inline data/startup.sh]: nit: it's more typical to have the first argument to mount be "none" or "tmpfs":

mount -t tmpfs tmpfs "${tmpdir}" or mount -t tmpfs none "${tmpdir}"

File: data/startup.sh [hunk 7]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+  # get OEMSize user input from meatadata
+  local -r oem_size="$(/usr/share/google/get_metadata_value \
+    attributes/OEMSize)"
+  cat > /etc/systemd/system/last-run.service<<EOF
+[Unit]
+Description=Run after everything unmounted
+DefaultDependencies=false
+Conflicts=shutdown.target
+Before=shutdown.target multi-user.target mnt-stateful_partition.mount var.mount mnt-disks.mount var-lib-docker.mount var-lib-toolbox.mount usr-share-oem.mount
```
Comment (rkolchmeyer) [inline data/startup.sh]: I think the only mounts you need here are "mnt-stateful_partition.mount" and "usr-share-oem.mount". "mnt-disks.mount" is just a tmpfs, and the other var mounts must be unmounted before mnt-stateful_partition.mount is unmounted.

Also, are shutdown.target and multi-user.target necessary here?

File: data/startup.sh [hunk 8]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+
+[Service]
+Type=oneshot
+RemainAfterExit=true
+ExecStart=/bin/true
+ExecStop=/tmp/last/extend-oem.bin /dev/sda 1 8 ${oem_size}
+TimeoutStopSec=600
+StandardOutput=tty
+StandardError=tty
+TTYPath=/dev/ttyS2
```
Comment (rkolchmeyer) [inline data/startup.sh]: Is it possible to prefix the extend-oem.bin stdout/stderr with BuildStatus? That will make the logs show in the Cloud Build logs, which will make debugging easier.

File: data/startup.sh [hunk 9]
```diff
@@ -141,6 +141,84 @@ fetch_builtin_ctx() {
...
+}
+
+extend_oem_partition(){
+  echo "Checking whether need to extend OEM partition..."
+
+  # get user input from meatadata
+  local -r oem_size="$(/usr/share/google/get_metadata_value \
+    attributes/OEMSize)"
+
+  if [ ${oem_size} != "" ]; then
```
Comment (rkolchmeyer) [inline data/startup.sh]: It's typically better to keep the core logic unindented, where possible:

if [[ -z "${oem_size}" ]]; then
  echo "No request to change OEM partition."
  return
fi
...

Comment (rkolchmeyer) [inline data/startup.sh]: nit: In bash, use '[[' instead of '['. Also, you can check if a variable is empty with "[[ -z "${oem_size}" ]]". You can check if a variable is non-empty with "[[ -n "${oem_size}" ]]".

Also, it's better to quote variables to prevent shell wildcard expansion: "${oem_size}".

File: preloader/preload.go [hunk 10]
```diff
@@ -66,6 +66,15 @@ func buildCloudConfig(script io.Reader, service io.Reader) (string, error) {
 		"systemctl daemon-reload",
 		"systemctl --no-block start customizer.service",
 	}
+
+	cloudConfig["bootcmd"] = []string{
+		"echo \"Resizing OEM partition file system...\"",
+		"umount /dev/sda8",
+		"e2fsck -fp /dev/sda8",
+		"resize2fs /dev/sda8",
+		"systemctl start usr-share-oem.mount",
```
Comment (rkolchmeyer) [inline preloader/preload.go]: It might be safer to put --no-block here, in case cloud-init somehow end up having a dependency relationship with usr-share-oem.mount.

File: preloader/preload.go [hunk 11]
```diff
@@ -66,6 +66,15 @@ func buildCloudConfig(script io.Reader, service io.Reader) (string, error) {
 		"systemctl daemon-reload",
 		"systemctl --no-block start customizer.service",
 	}
+
+	cloudConfig["bootcmd"] = []string{
```
Comment (rkolchmeyer) [inline preloader/preload.go]: Why not put this in the startup script?

File: preloader/preload.go [hunk 12]
```diff
@@ -143,19 +152,36 @@ func writeDaisyWorkflow(inputWorkflow string, outputImage *config.Image, buildSp
...
 	}
+
+	// template content for the step resize-disk. If disk-size is not provided
+	// a place holder is used, because ResizeDisk API requires a larger size than
+	// the original disk.
+	var resizeDiskJSON string
+	if buildSpec.DiskSize != 0 {
+		// actual disk size
+		resizeDiskJSON = fmt.Sprintf("\"ResizeDisks\": [{\"Name\": \"boot-disk\","+
+			"\"SizeGb\": \"%d\"}]", buildSpec.DiskSize)
```
Comment (rkolchmeyer) [inline preloader/preload.go]: You can use Go raw strings (backticks) to avoid escaping the quotes:

`"ResizeDisks": [{"Name": "boot-disk", "SizeGb": "%d"}]`

File: tools/main_for_build/extend_oem_bin.go [hunk 13]
```diff
@@ -0,0 +1,32 @@
+package main
```
Comment (rkolchmeyer) [inline tools/main_for_build/extend_oem_bin.go]: Instead of putting this file in a main_for_build directory, you can either put it directly in the tools/ directory or in a tools/cmd/ directory.

File: tools/main_for_build/extend_oem_bin.go [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"cos-customizer/tools"
+	"log"
+	"os"
+	"strconv"
+)
+
+// main generates binary file to extend the OEM partition
+// cmd: GOOS=linux GOARCH=amd64 CGO_ENABLE=0 go build -o ./../../data/builtin_build_context/extend-oem.bin extend_oem_bin.go
```
Comment (rkolchmeyer) [inline tools/main_for_build/extend_oem_bin.go]: Does this comment do anything? Ideally this should be built with Bazel.

File: tools/main_for_build/extend_oem_bin.go [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+func main() {
+	args := os.Args
+	if len(args) != 5 {
+		log.Println("error: must have 4 arguments: disk string, statePartNum, oemPartNum int, oemSize string")
+		return
```
Comment (rkolchmeyer) [inline tools/main_for_build/extend_oem_bin.go]: You can do log.Fatal for errors.

File: cmd/finish_image_build.go [hunk 16]
```diff
@@ -93,13 +97,42 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
 	flags.DurationVar(&f.timeout, "timeout", time.Hour, "Timeout value of the image build process. Must be formatted "+
 		"according to Golang's time.Duration string format.")
 }
 
 func (f *FinishImageBuild) validate() error {
+	// The default size of a COS image (imgSize) is 10GB. And the OEM partition size should be doubled to
+	// store the hash tree of dm-verity. If the OEM partition is to be extended,
+	// the extended disk size must be larger than oem-size x 2 + image size.
+	// Since we allow user input like "500M", the oem-size is rounded up to GB to make sure
+	// there is enough space. Extra space will be taken by the stateful partition.
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: Is it because extend disk size only takes GB as input? If yes, then can we add that too as why rounded upto nearest GB size is needed. If not, can you explain as why rounded upto GB is needed?

File: cmd/finish_image_build.go [hunk 17]
```diff
@@ -93,13 +97,42 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+	const imgSize = 10
+	if f.oemSize != "" {
+		oemSizeBytes, err := partutil.ConvertSizeToBytes(f.oemSize)
+		if err != nil {
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", f.oemSize, err)
+		}
+		f.oemFSSize4K = oemSizeBytes >> 12
+		// double the oem size.
+		oemSizeBytes <<= 1
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go:124]: Should we check integer overflow? I don't think it will be a realistic case though.

File: cmd/finish_image_build.go [hunk 18]
```diff
@@ -93,13 +97,42 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+		oemSizeGB, err := partutil.ConvertSizeToGBRoundUp(strconv.Itoa(oemSizeBytes) + "B")
+		if err != nil {
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", f.oemSize, err)
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go:127]: The actual value that is passed to above API is the value computed by ConvertSizeToBytes (which takes an user input). I think we should specify the bytes size here (though not sure how the format of that will be invalid).

File: cmd/finish_image_build.go [hunk 19]
```diff
@@ -93,13 +97,42 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+		}
+		if f.diskSize-oemSizeGB < imgSize {
+			return fmt.Errorf("'disk-size-gb' must be at least 'oem-size' x 2 + 10GB")
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: Should it be imgSize as you are checking against that.

File: cmd/finish_image_build.go [hunk 20]
```diff
@@ -93,13 +97,42 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+		}
+
+		// shrink OEM size input (rounded down) by 1M to deal with cases
+		// where disk size is 1M smaller than needed.
+		f.oemSize = strconv.Itoa((oemSizeBytes>>20)-1) + "M"
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: Is it to handle the case where you were getting 1MB less when the size was 12GB and getting the correct size when size was 20GB?

File: cmd/finish_image_build.go (hunk 19 above)
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go]: I think it needs to be clearly documented in cos customizer documentation as implicitly we assume that the disk size provided should satisfy `oemSize * 2 + 10GB`.

File: cmd/run_script.go [hunk 21]
```diff
@@ -100,7 +100,7 @@ func (r *RunScript) Execute(_ context.Context, f *flag.FlagSet, args ...interfac
 	}
 	files := args[0].(*fs.Files)
 	if r.script == "" {
-		log.Println("script not provided for %s step; script is required", r.Name())
+		log.Printf("script not provided for %s step; script is required\n", r.Name())
```
Comment (vaibhav-rustagi) [inline cmd/run_script.go:103]: Any reason for this?

File: data/startup.sh [hunk 22]
```diff
@@ -141,6 +141,90 @@ fetch_builtin_ctx() {
...
+StandardOutput=tty
+StandardError=tty
+TTYPath=/dev/ttyS2
+EOF
+}
+
+extend_oem_partition(){
+  echo "Checking whether need to extend OEM partition..."
+
+  # get user input from meatadata
```
Comment (vaibhav-rustagi) [inline data/startup.sh:189]: metadata

File: data/startup.sh [hunk 23]
```diff
@@ -141,6 +141,90 @@ fetch_builtin_ctx() {
...
+    mv builtin_ctx_dir/extend-oem.bin /tmp/extend-oem.bin
+    systemctl --no-block start last-run.service
+    stop_journald_service
+    echo "Rebooting..."
+    
+    # overwrite trap to avoid build failure triggered by reboot.
+    trap - EXIT
+    reboot
+    # keep it inside of this function until reboot kills the process
+    while :
```
Comment (vaibhav-rustagi) [inline data/startup.sh:221]: Can you explain this? Sorry couldn't get the reason for this logic.

File: preloader/preload.go [hunk 24]
```diff
@@ -143,19 +143,36 @@ func writeDaisyWorkflow(inputWorkflow string, outputImage *config.Image, buildSp
 		return "", err
 	}
 	acceleratorsJSON, err := json.Marshal([]map[string]interface{}{})
+	if err != nil {
+		return "", err
```
Comment (vaibhav-rustagi) [inline preloader/preload.go:147]: Should we add more context for the error OR the current error message information is sufficient to understand the behavior?

File: preloader/preload.go [hunk 25]
```diff
@@ -143,19 +143,36 @@ func writeDaisyWorkflow(inputWorkflow string, outputImage *config.Image, buildSp
...
 	if buildSpec.GPUType != "" {
 		acceleratorType := fmt.Sprintf("projects/%s/zones/%s/acceleratorTypes/%s",
 			buildSpec.Project, buildSpec.Zone, buildSpec.GPUType)
 		acceleratorsJSON, err = json.Marshal([]map[string]interface{}{
 			{"acceleratorType": acceleratorType, "acceleratorCount": 1}})
-	}
-	if err != nil {
-		return "", err
+		if err != nil {
+			return "", err
```
Comment (vaibhav-rustagi) [inline preloader/preload.go:155]: Same as above.

File: tools/cmd/extend_oem/extend_oem_bin.go [hunk 26]
```diff
@@ -0,0 +1,30 @@
+package main
```
Comment (vaibhav-rustagi) [inline tools/cmd/extend_oem/extend_oem_bin.go:15]: Any reason for naming the file: extend_oem_bin.go and not extend_oem.go?

File: cmd/finish_image_build.go (hunk 20 above)
Comment (hegao12) [reply cmd/finish_image_build.go]: Yes it is.

File: cmd/run_script.go (hunk 21 above)
Comment (hegao12) [reply cmd/run_script.go:103]: there is a `%s` inside of the string.

File: preloader/preload.go (hunk 24 above)
Comment (hegao12) [reply preloader/preload.go:147]: This is not part of my code and I don't know what the error message should be.  I modified it just because I found that one error check was missing.

File: tools/cmd/extend_oem/extend_oem_bin.go (hunk 26 above)
Comment (hegao12) [reply tools/cmd/extend_oem/extend_oem_bin.go:15]: It might be confused with tools/extend_oem_partition.go.
The "_bin" shows that this Go program is used to generate a binary.

File: preloader/preload.go (hunk 24 above)
Comment (vaibhav-rustagi) [reply preloader/preload.go:147]: SGTM

**Comment by vaibhav-rustagi** [comment]
data/startup.sh is showing a conflict. Kindly resolve it.

File: BUILD.bazel [hunk 27]
```diff
@@ -49,6 +50,20 @@ genrule(
...
+genrule(
+    name = "copy_extend_oem_bin",
+    srcs = ["//tools/cmd/extend_oem:main_for_build"],
+    outs = ["extend-oem.bin"],
+    cmd = "mv $< $@",
+)
+
+pkg_tar(
+    name = "remap_extend_oem_bin",
+    srcs = ["copy_extend_oem_bin"],
```
Comment (rkolchmeyer) [inline BUILD.bazel]: I'm not sure why this works; but "copy_extend_oem_bin" should instead be ":copy_extend_oem_bin", since you are referencing the output of another target.

File: BUILD.bazel [hunk 28]
```diff
@@ -49,6 +50,20 @@ genrule(
     cmd = "mkdir $@",
 )
 
+genrule(
+    name = "copy_extend_oem_bin",
+    srcs = ["//tools/cmd/extend_oem:main_for_build"],
+    outs = ["extend-oem.bin"],
+    cmd = "mv $< $@",
```
Comment (rkolchmeyer) [inline BUILD.bazel]: nit: is there a specific reason why this is "mv" and not "cp"? I'm not sure if genrules like it when the inputs are modified.

File: BUILD.bazel [hunk 29]
```diff
@@ -77,3 +93,9 @@ go_image(
     pure = "on",
     visibility = ["//visibility:public"],
 )
+
+go_binary(
+    name = "cos-customizer",
+    embed = [":go_default_library"],
+    visibility = ["//visibility:public"],
+)
```
Comment (rkolchmeyer) [inline BUILD.bazel]: I know that gazelle generates this rule, but we don't need it. So let's remove it.

File: cmd/finish_image_build.go [hunk 30]
```diff
@@ -93,13 +97,44 @@ func (f *FinishImageBuild) SetFlags(flags *flag.FlagSet) {
...
+	// Extra space will be taken by the stateful partition.
+	const imgSize = 10
+	if f.oemSize != "" {
+		oemSizeBytes, err := partutil.ConvertSizeToBytes(f.oemSize)
+		if err != nil {
+			return fmt.Errorf("invalid format of oem-size: %q, error msg:(%v)", f.oemSize, err)
+		}
+		f.oemFSSize4K = oemSizeBytes >> 12
+		// double the oem size.
+		oemSizeBytes <<= 1
```
Comment (rkolchmeyer) [inline cmd/finish_image_build.go:124]: Let's leave the dm-verity stuff out of this PR. We can have a separate PR for incorporating all of the dm-verity work.

File: WORKSPACE [hunk 31]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 Google LLC
+# Copyright 2020 Google LLC
```
Comment (rkolchmeyer) [inline WORKSPACE]: It looks like a recent Bazel upgrade broke cos-customizer builds, and these WORKSPACE changes are needed for cos-customizer to compile again. This makes these WORKSPACE changes much higher priority than the rest of the changes in this PR. Can we separate out these WORKSPACE changes into a different PR, and we can submit that?

File: data/startup.sh [hunk 32]
```diff
@@ -47,7 +47,7 @@ exit_workdir() {
 setup() {
   echo "Setting up the environment for preloading..."
   if systemctl status update-engine; then
-    systemctl stop update-engine
+    systemctl stop update-engine || :
```
Comment (rkolchmeyer) [inline data/startup.sh]: A "stop_service" function was added recently; you might be able to use that here.

File: testing/extend_oem_test/preload.sh [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+  docker pull ubuntu && break || docker_code="$?"
+  i=$((i+1))
+done
+if [[ $i -eq 11 ]]; then
+  echo "Pulling ubuntu failed."
+  echo "Docker journal logs:"
+  journalctl -u docker.service --no-pager
+  exit "${docker_code}"
+fi
+echo "Successfully pulled ubuntu container image."
```
Comment (rkolchmeyer) [inline testing/extend_oem_test/preload.sh:32]: Consider writing something to the OEM partition in this script as well.

File: tools/cmd/extend_oem/BUILD.bazel [hunk 34]
```diff
@@ -0,0 +1,15 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
```
Comment (rkolchmeyer) [inline tools/cmd/extend_oem/BUILD.bazel:15]: I think you need to add a license header here

File: tools/cmd/extend_oem/extend_oem_bin.go (hunk 26 above)
Comment (rkolchmeyer) [inline tools/cmd/extend_oem/extend_oem_bin.go:15]: I think you need a license header here too

File: cmd/finish_image_build.go (hunk 30 above)
Comment (rkolchmeyer) [reply cmd/finish_image_build.go:124]: Discussed offline, we'll keep this in for now and improve in follow-up PR.

File: data/startup.sh [hunk 35]
```diff
@@ -47,7 +47,7 @@ exit_workdir() {
 setup() {
   echo "Setting up the environment for preloading..."
   if systemctl status update-engine; then
-    systemctl stop update-engine
+    stop_service update-engine
   else
     echo "'systemctl status update-engine' failed; this is non-fatal"
   fi
```
Comment (rkolchmeyer) [inline data/startup.sh]: I think this whole if block can be replaced with `stop_service update-engine`

**Review by rkolchmeyer — APPROVED** [review]
